lib/chromatin-data.js

```diff
@@ -0,0 +1,482 @@
+  "Active enhancer 1",
+  "Active enhancer 2",
+  "Weak enhancer",
+  "Genic enhancer 1",
+  "Genic enhancer 2",
+  "Strong transcription",
+  "Weak transcription",
+  "Bivalent/Poised TSS",
+  "Bivalent enhancer",
+  "ZNF genes & repeats",
+  "Weak Repressed PolyComb",
+  "Repressed PolyComb",
+  "Heterochromatin",
+  "Quiescent/Low",
+];
+
+export function isLetter(c) {
+  return c.toLowerCase() !== c.toUpperCase();
+}
+
+/**
+ * It filters and cleans up the original data and returns a cleaned data array.
+ * @param {Array} data The original data got from GDS
+ * @returns The cleaned data array
+ */
+export function getChromatinData(data) {
+  const filteredData = data.filter((d) => d.method === "chromatin state");
+  const chromatinData = [];
+  if (filteredData.length > 1) {
+    for (let i = 0; i < filteredData.length; i++) {
+      const dataset = {};
+      const chromatinState = !isLetter(filteredData[i].value[0])
+        ? filteredData[i].value.replace(/[^A-Za-z]+/g, "")
+        : filteredData[i].value;
+      dataset.chromatin_state = mapChromatinNames[chromatinState];
+      dataset.organ =
+        filteredData[i].biosample_ontology?.organ_slims?.join(", ") || "";
+      dataset.biosample = filteredData[i].biosample_ontology
+        ? filteredData[i].biosample_ontology.term_name
+        : "";
+      dataset.biosample_classification = filteredData[i].biosample_ontology
+        ? filteredData[i].biosample_ontology.classification
+        : "";
+      dataset.dataset = filteredData[i].dataset;
+      dataset.file = filteredData[i].file;
+      chromatinData[i] = dataset;
+    }
+  }
+  return chromatinData;
+}
+
+/**
+ * It generates the facets data used for BodyMap.
+ * The data is a object contains keys of different organ slims.
+ * for each organ slim key, the value is an array of the states the organ slim has,
+ * sorded by state activity.
+ * @param {array} data to get organ facets
+ * @param {string} assembly for the data
+ * @returns the organ facets data
+ */
+export function getOrganFacets(data, assembly) {
+  const order =
+    assembly === "hg19"
+      ? SORTED_CHROMATIN_STATES_HG19
+      : SORTED_CHROMATIN_STATES_GRCH38;
+  const uniqueOrgansWithSates = {};
+  data.forEach((dataset) => {
+    const organs = dataset.organ.split(", ").filter((e) => {
+      return e;
+    });
+    const state = dataset.chromatin_state;
+    organs.forEach((organ) => {
+      if (organ in uniqueOrgansWithSates) {
+        if (!uniqueOrgansWithSates[organ].includes(state)) {
+          uniqueOrgansWithSates[organ].push(state);
+        }
+      } else {
+        uniqueOrgansWithSates[organ] = [state];
+      }
+    });
+  });
+  const sorted = {};
+  Object.keys(uniqueOrgansWithSates).forEach((organ) => {
+    const sortedStates = [];
+    order.forEach((state) => {
+      if (uniqueOrgansWithSates[organ].includes(state)) {
+        sortedStates.push(state);
+      }
+    });
+    sorted[organ] = sortedStates;
+  });
```

Comment:
I think all these lops can be replaced by `.map()`. I just find that more obvious when the input array and output array have the same number of items, guaranteed.